addons/mail/models/mail_thread.py

```diff
@@ -1762,6 +1762,13 @@ def _mail_find_partner_from_emails(self, emails, records=None, force_create=Fals
             followers = records.mapped('message_partner_ids')
         else:
             followers = self.env['res.partner']
+        if records and records._fields.get('company_id'):
+            sort_key = lambda p: (
+                self.env.user.partner_id == p,
+                p.company_id <= records.company_id,
+                p.commercial_partner_id <= records.company_id.partner_id)
```

Comment:
so this line will associate a commercial partner to the res.company().partner_id

Usually a partner associated to an existing `res.company` will be an employee that has a res.users that has unicity over the login. I guess this could improve the case where someone has a second res.partner with the same email as his res.user has duplicated partner, but I think that if `_mail_search_on_user` (that have unicity) find a user, we won't search partner for him.

So not sure what this line fixes but I could miss something.

Reply:
Yes, that's a valid point. Added it because that's how I initially tested it, but it is an unlikely scenario, and not required to fix the client's issue.